reference/matrix/diagonal_kernels.cpp

```diff
@@ -132,6 +132,24 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_DIAGONAL_RIGHT_APPLY_TO_CSR_KERNEL);
 
 
+template <typename ValueType, typename IndexType>
+void fill_in_matrix_data(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const Array<matrix_data_entry<ValueType, IndexType>>& data,
+    matrix::Diagonal<ValueType>* output)
+{
+    for (size_type i = 0; i < data.get_num_elems(); i++) {
+        const auto entry = data.get_const_data()[i];
+        if (entry.row == entry.column) {
+            output->get_values()[entry.row] = entry.value;
```

Comment:
It sounds like you're describing invalid matrix_data with out-of-bounds entries - should that be supported/caught? I'd probably defer that to device_matrix_data::validate() and #770

Reply:
Maybe? it something like you did for core/test/utils/matrix_generator.hpp?

Reply:
That was a bugfix for something that was broken earlier. To be honest, I think we should not have added a `read` function to Diagonal at all, since it is not a generic matrix format. But that was an earlier oversight, we can't change that until 2.0. Until then, I'd like to avoid having deal with device-side error handling